File: script/models/AccountTemplate.py
from peewee import *
from playhouse.postgres_ext import BinaryJSONField
from .BaseWithTimeZoneModel import BaseWithTimeZoneModel
from .Account import Account
from .Template import Template


class AccountTemplate(BaseWithTimeZoneModel):
    """
    Pivot model linking accounts to the templates they have posted.
    created_at is inherited from BaseWithTimeZoneModel and acts as both
    the assignment time (for pending rows) and the post time (the worker
    overwrites it with the real post timestamp on completion).
    """
    account = ForeignKeyField(Account, backref='templates')
    template = ForeignKeyField(Template, backref='account_template')

    status = CharField(default='pending')
    url = CharField(null=True)

    like_count = IntegerField(default=0)
    comment_count = IntegerField(default=0)
    view_count = IntegerField(default=0)
    save_count = IntegerField(default=0)
    repost_count = IntegerField(default=0)

    stats = BinaryJSONField(null=True)

    updated_at = DateTimeField(null=True)

    statuses = [
        'pending',
        'processing',
        'completed',
    ]

    class Meta:
        table_name = 'account_template'

File: backend/app/Models/AccountTemplate.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Relations\Pivot;

class AccountTemplate extends Pivot
{
    protected $table = 'account_template';

    public $incrementing = true;

    // We manage timestamps manually. created_at is set by the worker
    // when a post completes (and by the assign endpoint when a row is
    // created in 'pending' state). updated_at is bumped when stats or
    // status are refreshed.
    public $timestamps = false;

    protected $guarded = [];

    public static array $statuses = [
        'pending',
        'processing',
        'completed',
    ];

    protected $casts = [
        'stats'         => 'array',
        'created_at'    => 'datetime',
        'updated_at'    => 'datetime',
        'like_count'    => 'integer',
        'comment_count' => 'integer',
        'view_count'    => 'integer',
        'save_count'    => 'integer',
        'repost_count'  => 'integer',
    ];

    public function account()
    {
        return $this->belongsTo(Account::class);
    }

    public function template()
    {
        return $this->belongsTo(Template::class);
    }
}
